struct_ui rework

phatchman opened the pull request:
Implements #488 

## Description ##

Overhaul of the struct_ui functionality. Highlights are:
1) Field options are now per type, rather than per field. This simplifies the options processing considerably.
2) Decomposed the API into ~3 levels of publicly accessible functions
    - Individual widget display functions (e.g. numberFieldWidget)
    - Field display functions that call one or more widget display functions ( e.g. displayNumber)
    - Container display functions to display a struct or union. (e.g. displayStruct)
3) Improved compile-time errors messages.
4) Renamed from structUI to struct_ui and generally conform to zig naming standards.
5) Adds a "depth" parameter to the struct display, so that the user can control how many levels of nested structs to display. A depth of 0 will show all non-struct fields, including pointers, optionals or unions. Depth is only incremented when processing a struct fields.

Hopefully the new comptime code is simpler to understand and easier to use.
 
 ## Discussion / TODO ##
 
 This is really just to get something out there showing the reworked API and some usage examples. There are still a few things to clean up, but no changes to the API are expected.
 
1. I ended up deleting the data adapter changes (deja vu). They do work quite well and are useful, but:
   - I'm not sure how much use struct_ui will ultimately get and whether it is worth adding complexity until we know there is a need for it.
   - Just from a usability standpoint, it adds another comptime path that the user needs to get right, requires passing pointers to adapters, rather than copies of adapters, which all needs comptime checking with nice error messages to make it manageable. i.e. it's a surprising amount of code to add for just being able to modify a string. (well not just that, but mostly that).
   - If there is a need, it is easy to bring it back, maybe in a simpler form. I'll keep the branch alive.
2. As a result, strings are handled in a bit of a hacky, but simple way. They get assigned to the TextWidget's internal dynamically allocated buffer. If the struct should live longer than the lifetime of the window (or outside the window begin/end??), the user will have to dupe the strings. Hopefully that is an OK compromise for now.
3. The theme editor in demos works using the new APIs.
4. ~~I've temporarily hijacked sdl-standalone for some struct tests.~~
5. ~~There are some things that still need fixing, like optionals for strings, but it's working for all struct combinations I've managed to think up so far.~~
6. ~~There is a lot more comptime error checking to be done. I've found you need to check the `type` and `anytype` parameters at each level to get any sort of comptime debuggability.~~
7. ~~The UI needs a bit of a cleanup for alignment and display of optionals and unions. But I'm trying to keep it to a fairly minimal look based around expanders grouping together fields in a container.~~
8. Probably won't expose any of it via dvui.zig functions. People can just import the dvui.struct_ui namespace and use the functions directly. [Note now exposes one function named structUI, to make sure the displayStruct is wrapped in a box to ensure a unique WidgetId.]
10. Potentially make changes to the options so fields can be totally skipped, rather than just "not displayed". - Done
11. ~~displayStruct doesn't take @src(), so if someone does a displayStruct on the same type more than once, they will get duplicate id errors. So maybe it is worth exposing just his one via a dvui.zig function to wrap the call in a box. I could make it take @src(), and pass it everywhere, but after creating the first box, it is no longer needed, everything else will get a unique src based on the comptime expansion for each type.~~
12. Add an example showing an advanced use of the API, e.g. for an optional pointer to a union containing a struct or similar with 3 levels of nesting, with the custom handling for the optional pointer in the 2nd level.
13. ~~Implement read-only display for all widget functions.~~

## Theme editor 
<img width="858" height="962" alt="image" src="https://github.com/user-attachments/assets/dbd42ad0-fd3d-4345-b142-5b942cf82799" />

david-vanderson commented:
Excellent!  Will dig into it!

phatchman commented:
Can anyone thing of a good way of testing comptime error messages? 

I need to exercise all the error paths at least once. Outside of just checking the errors look OK, there could be syntax errors, bugs etc in the error handling paths that don't get found until that code is run during compile time.

Right now I have a fn that contains lots of calls with invalid parameters and comment out all the lines except the current one I'm testing. But that's not really sustainable outside of a 1-off test.

I can think of complicated ways to do it using conditional compilation and the build system, but that seems like a huge overkill to test some error messages. I need comptime tests or something.

david-vanderson commented:
This is looking really good!

> Also let me know if you see any issues with the way I'm handling the strings with assigning the slices to the string widget's buffer. There is an alternative displayStringBuf() that can be used if the user wants to manually pass in a buffer for the text widget to use instead.

That is a good idea, and I like pushing the ownership of the string back onto the usage code.  Maybe I missed it, but how would the usage code know if it needed to dupe/deal with a changed string?

phatchman commented:
What do you think about something like the below: (Yes, I'll try and come up with a better name for this)
```
var sw: StringWrangler = .{};
```

```
fn gui_frame() void {
            if (!sw.initialized) {
                sw.init(&top_string.top_str);
            }
            dvui.structUI(@src(), "top_string", &top_string, 2, .{ top_options, base_options });
            if (sw.changed()) {
                std.debug.print("cchanged\n", .{});
                sw.wrangle(gpa) catch |err| @panic(@errorName(err));
            }
}
```
For each string, you create a "Wrangler" that:
1) Keeps track of the original pointer, so knows if it has changed.
2) Knows if the string is statically or dynamically allocated (user will also be able to set this during init())
3) Will dup the string only if it has changed.
4) Has a deinit() so if the string was dynamically allocated, it gets freed at the end.

In the example above, you probably wouldn't "wrangle" on each change, only when the window is closed. The wrangler needs to have the same lifetime as the struct, so when you open the window again, you'd use the same wranglers from previous to retain the state about whether the strings are dynamically allocated.

```
pub const StringWrangler = struct {
    initialized: bool = false,
    dynamically_allocated: bool = false,
    str_ptr: ?*[]const u8 = null,
    str_slice: ?[]const u8 = null,

    pub fn init(self: *StringWrangler, str_ptr: *[]const u8) void {
        self.initialized = true;
        self.str_ptr = str_ptr;
        self.str_slice = str_ptr.*;
    }

    pub fn changed(self: StringWrangler) bool {
        if (self.initialized) {
            return self.str_ptr.?.*.ptr != self.str_slice.?.ptr;
        } else {
            return false;
        }
    }

    pub fn wrangle(self: *StringWrangler, allocator: std.mem.Allocator) !void {
        if (self.initialized and self.changed()) {
            if (self.dynamically_allocated) {
                allocator.free(self.str_slice.?);
            }
            self.str_slice = try allocator.dupe(u8, self.str_ptr.?.*);
            self.dynamically_allocated = true;
        }
    }

    pub fn deinit(self: *StringWrangler, allocator: std.mem.Allocator) void {
        if (self.initialized and self.dynamically_allocated) {
            allocator.free(self.str_slice.?);
        }
    }
};
```

It's basically the "remember the original pointer address" solution, but with some helpers to make it nicer.

david-vanderson commented:
Wow this is a hard problem.  It feels like we're trying to do too much automatically with not enough information.

Maybe we should push reallocating strings into a future problem.  Right now we can show read-only strings, and read-write in-place strings right?

I think this structUI stuff is very useful and don't want to hold you up, so I say let's merge this and punt on heap reallocation for now.

What do you think?

david-vanderson commented:
> Unless you can see a better way of handling var strings that doesn't require changing what the slice points to, I'd probably merge what we have today and then refine the string wrangler/handler when I do the PR to add some more advanced demos.

Nope.  Let's merge this!

I very much appreciate you laying out all the different approaches.

> See [this example](https://github.com/phatchman/dvui-dev/blob/f4a32affdebed9d242db7f98e2ba5e007dff4a58/src/Examples/struct_ui.zig) for a simple string handling demo where the strings can outlast the lifetime of the window.

I like how that is looking, need to spend more time playing with it.  I don't quite have the logic about when to set `str_ptr` vs `str_slice` straight in my head.

> There is also some bug with the deinit / init in the demo that I still need to work out.

One thing that has gotten me a few times is that all the demo code runs twice - once for the button you click on, and again for the real thing, so any globals might be being fought over.  I've been trying to slowly change them to all use dataGet/dataSet to deal with this.

david-vanderson commented:
> As far as I understand, the buffer backing text widget won't be deallocated until deinit() is called on the Window.

The buffer is allocated using dataGet/dataSet right?  That stuff should go away as soon as there is a frame where it's not referenced (the id/name combo is not used in any dataGet or dataSet).  Are you not seeing that?

> Maybe the string duplication really is a niche case?

I think this is probably the case.  Especially for a general-purpose struct viewer.

Given the enormous difficulty we've had so far with memory allocation, I suspect a solution might involve making some tradeoff, like saying "editing slices with structUI is going to leak memory" (so okay for development/one-off stuff).

I very much appreciate the time and effort you've put into this - thank you!

phatchman commented:
> The buffer is allocated using dataGet/dataSet right? That stuff should go away as soon as there is a frame where it's not referenced (the id/name combo is not used in any dataGet or dataSet). Are you not seeing that?
> 

Yeah, I am. 1) I missed the code that does the freeing and 2) I needed a more elaborate test-case to cause it to fail. 

That means anything that is not always displayed (e.g. anything in an expander) has the same problem, which means all strings will have the problem because I display structs in expanders.

Hmm. OK. I think it needs a better solution then.
